### PR TITLE
Feat: Support basic-auth or reuse chartmuseum as addon registry

### LINKED ISSUE #3618: [Feature] Support basic-auth or reuse chartmuseum as  addon registry
**Is your feature request related to a problem? Please describe.**

Origin from https://github.com/oam-dev/kubevela/pull/3595#discussion_r845755238

We have support many types addon registry such as git. OSS and helm repo. Helm repo can be used as a versioned registry.
We should support private helm repo  which setup with basic-auth(need set http username and password)

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by StevenLeiZhang**
Signed-off-by: StevenLeiZhang <zhangleiic@163.com>

### Description of your changes

Fixes #3618 

1. add one vela command `vela addon package <addon directory>` to package vela addon directory into a helm chart compatible archive
example:
```
➜  ./vela addon package ./velaux
Successfully package addon to: /KubeVela/addons/velaux-v1.3.3.tgz
```

package command will create one "Chart.yaml" file, and its content is filled by "metadata.yaml"

```
➜  addons tree ./velaux
./velaux
├── Chart.yaml
├── metadata.yaml
├── readme.md
├── readme_cn.md
├── resources
│   ├── apiserver.cue
│   ├── parameter.cue
│   ├── velaux.cue
│   └── views.cue
├── schemas
│   ├── addon-uischema-velaux.yaml
│   ├── component-uischema-k8s-objects.yaml
│   ├── component-uischema-task.yaml
│   ├── component-uischema-webservice.yaml
│   ├── trait-uischema-scaler.yaml
│   ├── trait-uischema-storage.yaml
│   ├── workflowstep-uischema-deploy2env.yaml
│   ├── workflowstep-uischema-notification.yaml
│   └── workflowstep-uischema-webhook.yaml
└── template.yaml

2 directories, 18 files
➜  addons cat ./velaux/Chart.yaml
apiVersion: v2
appVersion: v1.3.3
description: KubeVela User Experience (UX). An extensible, application-oriented delivery
  and management Dashboard.
home: https://kubevela.io
icon: https://static.kubevela.net/images/logos/KubeVela%20-03.png
keywords:
- Official
name: velaux
type: library
version: v1.3.3
```

Please note that: before "package", this directory must be one valid vela addon with all mandatory content. 

I have followed the two documents to mapp "matadata.yaml" and "Chart.yaml"
https://kubevela.net/docs/next/platform-engineers/addon/intro
https://helm.sh/docs/topics/charts/

|  vela   | helm  |
|  ----  | ----  |
| name  | name |
| description  | description |
| version | version |
| version | appVersion |
| icon | icon |
| url | home |
| tags | keywords |

2. add 2 arguments for "vela addon registry add" to specify username and password
example:

```
➜  bin git:(fea-auth) ✗ ./vela addon registry add harbor --endpoint="https://hub.harbor.com/chartrepo/oam-addon" --username=xxxxxxxx --password=xxxxxxx --type=helm
Successfully add an addon registry harbor 
```

I have:

- [ ] Read and followed KubeVela's [contribution process](https://github.com/oam-dev/kubevela/blob/master/contribute/create-pull-request.md).
- [ ] [Related Docs](https://github.com/oam-dev/kubevela.io) updated properly. In a new feature or configuration option, an update to the documentation is necessary. 
- [ ] Run `make reviewable` to ensure this PR is ready for review.
- [ ] Added `backport release-x.y` labels to auto-backport this PR if necessary.

### How has this code been tested

0.install Harbor to be private registry, or you could also use official demo site https://goharbor.io/docs/2.3.0/install-config/demo-server/
1. install KubeVela v1.3.1
4. delete default addon registry
```
➜  bin git:(fea-auth) ✗ ./vela addon registry delete KubeVela
Successfully delete an addon registry KubeVela 
```
5.upload fluxcd addon and velaux addon to Harbor
please refer to https://github.com/oam-dev/kubevela/issues/3618#issuecomment-1110794857 for details

6.add Harbor to be addon registry 
```
➜  bin git:(fea-auth) ✗ ./vela addon registry add harbor --endpoint="https://hub.harbor.com/chartrepo/oam-addon" --username=xxxxxxxx --password=xxxxxxx --type=helm
Successfully add an addon registry harbor 
```

7.list Harbor
```
➜  bin git:(fea-auth) ✗ ./vela addon registry list                                                                                                                             
Name    Type    URL                                          
harbor  helm    https://hub.harbor.com/chartrepo/oam-addon
➜  bin git:(fea-auth) ✗ ./vela addon list                                                                                                    
NAME    REGISTRY        DESCRIPTION                     AVAILABLE-VERSIONS      STATUS  
fluxcd  harbor          A kubevela addon for fluxcd     [1.1.0, 1.0.0]          disabled
velaux  harbor          A kubevela addon for velaux     [v1.3.1, v1.3.0]        disabled
```

8.enable velaux from Harbor
```
➜  bin git:(fea-auth) ✗  ./vela addon enable velaux --version=v1.3.1  
I0428 18:20:30.802898   61217 apply.go:107] "creating object" name="addon-uischema-velaux" resource="/v1, Kind=ConfigMap"
I0428 18:20:30.816572   61217 apply.go:107] "creating object" name="component-uischema-k8s-objects" resource="/v1, Kind=ConfigMap"
I0428 18:20:30.826370   61217 apply.go:107] "creating object" name="component-uischema-task" resource="/v1, Kind=ConfigMap"
I0428 18:20:30.838713   61217 apply.go:107] "creating object" name="component-uischema-webservice" resource="/v1, Kind=ConfigMap"
I0428 18:20:30.850772   61217 apply.go:107] "creating object" name="trait-uischema-scaler" resource="/v1, Kind=ConfigMap"
I0428 18:20:30.861949   61217 apply.go:107] "creating object" name="trait-uischema-storage" resource="/v1, Kind=ConfigMap"
I0428 18:20:30.876154   61217 apply.go:107] "creating object" name="workflowstep-uischema-deploy2env" resource="/v1, Kind=ConfigMap"
I0428 18:20:30.903761   61217 apply.go:107] "creating object" name="workflowstep-uischema-notification" resource="/v1, Kind=ConfigMap"
I0428 18:20:30.916231   61217 apply.go:107] "creating object" name="workflowstep-uischema-webhook" resource="/v1, Kind=ConfigMap"
I0428 18:20:30.927497   61217 apply.go:107] "creating object" name="addon-secret-velaux" resource="/v1, Kind=Secret"                                                                                                 Addon: velaux enabled Successfully.
Please access the velaux from the following endpoints:
+---------+-----------+----------------------------+---------------------------+
| CLUSTER | COMPONENT |  REF(KIND/NAMESPACE/NAME)  |         ENDPOINT          |
+---------+-----------+----------------------------+---------------------------+
|         | velaux    | Service/vela-system/velaux | http://192.168.65.4:31788 |
+---------+-----------+----------------------------+---------------------------+
```

9.check velaux status, and it should be enabled
```
➜  bin git:(fea-auth) ✗ ./vela addon list                                                                                                    
NAME    REGISTRY        DESCRIPTION                     AVAILABLE-VERSIONS      STATUS          
fluxcd  harbor          A kubevela addon for fluxcd     [1.1.0, 1.0.0]          disabled        
velaux  harbor          A kubevela addon for velaux     [v1.3.1, v1.3.0]        enabled (v1.3.1)
```

10.go to velaux webpage->addons to list all addons
image

11. click the card of fluxcd, go to its detail page and click button enable to enable fluxcd. Fluxcd addon should be enabled successfully.
image

12.check Fluxcd addon status , it should be enabled
```
➜  bin git:(fea-auth) ✗ ./vela addon list
NAME    REGISTRY        DESCRIPTION                     AVAILABLE-VERSIONS      STATUS          
fluxcd  harbor          A kubevela addon for fluxcd     [1.1.0, 1.0.0]          enabled (1.1.0) 
velaux  harbor          A kubevela addon for velaux     [v1.3.1, v1.3.0]        enabled (v1.3.1)
```

13. try to disable fluxcd. It should be disabled successfully
```
➜  bin git:(fea-auth) ✗ ./vela addon disable fluxcd        
Successfully disable addon:fluxcd
```

14. delete Harbor addon registy, it should be deleted successfully
```
➜  bin git:(fea-auth) ✗ ./vela addon registry delete harbor
Successfully delete an addon registry harbor 
```

15. Add KubeVela official registry back, and enable Fluxcd.It should works well.
```
➜  bin git:(fea-auth) ✗ ./vela addon registry add KubeVela --endpoint="https://addons.kubevela.net"  --type=helm
Successfully add an addon registry KubeVela 
➜  bin git:(fea-auth) ✗ 
➜  bin git:(fea-auth) ✗ ./vela addon registry list
Name            Type    URL                        
KubeVela        helm    https://addons.kubevela.net
➜  bin git:(fea-auth) ✗  ./vela addon enable fluxcd
➜  bin git:(fea-auth) ✗ ./vela addon list | grep flux
fluxcd                          KubeVela        Extended workload to do continuous and progressive delivery                                             [1.1.0, 1.0.0]                  enabled (1.1.0) 
➜  bin git:(fea-auth) ✗ 

```

### CODE REVIEW

**Comment by wonderflow** [comment]
I  think this PR hasn't fully fixed the issue https://github.com/kubevela/kubevela/issues/3618 , it's part of it. We need to generate chart.yaml automatically for the rest part.

File: references/cli/addon-registry.go [hunk 1]
```diff
@@ -277,6 +279,8 @@ func parseArgsFromFlag(cmd *cobra.Command) {
 	cmd.Flags().StringP(addonOssBucket, "", "", "specify the OSS bucket name")
 	cmd.Flags().StringP(addonPath, "", "", "specify the addon registry OSS path")
 	cmd.Flags().StringP(addonGitToken, "", "", "specify the github repo token")
+	cmd.Flags().StringP(addonUsername, "", "", "specify the addon registry username")
+	cmd.Flags().StringP(addonPassword, "", "", "specify the addon registry password")
```
Comment (wangyikewxgm) [inline references/cli/addon-registry.go:283]: Please point out  clearly the username and password args only can be used for helm repo registry.

**Comment by wangyikewxgm** [comment]
The code is generally LGTM. But I'm curious about how the `index.yaml` in harbor be generated? Do you write it manually?

**Comment by StevenLeiZhang** [comment]
> I think this PR hasn't fully fixed the issue #3618 , it's part of it. We need to generate chart.yaml automatically for the rest part.

Ok, Let me add `vela addon package` command to make this PR to be complete implementation of #3618

**Comment by StevenLeiZhang** [comment]
> The code is generally LGTM. But I'm curious about how the `index.yaml` in harbor be generated? Do you write it manually?

Ok, Let me add `vela addon package` command to make this PR to be complete implementation of #3618

File: pkg/addon/addon.go [hunk 2]
```diff
@@ -1438,3 +1445,51 @@ func fetchVelaCoreImageTag(ctx context.Context, k8sClient client.Client) (string
...
+		Name:        meta.Name,
+		Description: meta.Description,
+		// define Vela addon's type to be library in order to prevent installation of a common chart. Please refer to https://helm.sh/docs/topics/library_charts/
+		Type:       "library",
+		Version:    meta.Version,
+		AppVersion: meta.Version,
+		APIVersion: chart.APIVersionV2,
+		Icon:       meta.Icon,
+		Home:       meta.URL,
+		Keywords:   meta.Tags,
```
Comment (wonderflow) [inline pkg/addon/addon.go:1472]: what's the convert relationship between `metatdata.yaml` and `chart.yaml`? Please make sure we didn't miss any useful information here. \cc @wangyikewxgm

Comment (wonderflow) [reply pkg/addon/addon.go:1472]: we need to add a doc into `kubevela.io` to describe these details for guiding addon developers.

Comment (StevenLeiZhang) [reply pkg/addon/addon.go:1472]: > what's the convert relationship between `metatdata.yaml` and `chart.yaml`? Please make sure we didn't miss any useful information here. \cc @wangyikewxgm

Have updated description of this PR to explain the mapping. @wangyikewxgm pls review it.

Comment (StevenLeiZhang) [reply pkg/addon/addon.go:1472]: > we need to add a doc into `kubevela.io` to describe these details for guiding addon developers.

OK. I will update document later.

File: pkg/addon/addon.go [hunk 3]
```diff
@@ -1438,3 +1445,51 @@ func fetchVelaCoreImageTag(ctx context.Context, k8sClient client.Client) (string
...
+	ch, err := loader.LoadDir(addonDictPath)
+	if err != nil {
+		return "", err
+	}
+
+	dest, err := os.Getwd()
+	if err != nil {
+		return "", err
+	}
+	archive, err := chartutil.Save(ch, dest)
```
Comment (wangyikewxgm) [inline pkg/addon/addon.go:1490]: Is the package's name is {addonName}-{version}.tgz ?

File: pkg/addon/addon.go (hunk 2 above)
Comment (wangyikewxgm) [reply pkg/addon/addon.go:1472]: LGTM, the translate rule is totally same with the logic in catalog. https://github.com/kubevela/catalog/blob/4b46deb768b887e809934049f5aad6cefaa32a8b/hack/addons/syn_addon_package.go#L108

Comment (wangyikewxgm) [reply pkg/addon/addon.go:1472]: Please make this info clear in the doc. This file is only be used for generate the index.yaml for chart-museum automatically in "package command". This file is useless when enable an addon. So the addon's maker  needn't take aware of the exist of this file and mustn't write this file.

Comment (wonderflow) [reply pkg/addon/addon.go:1472]: the CI in catalog should use the same code with this PR after merged @wangyikewxgm

Comment (wangyikewxgm) [reply pkg/addon/addon.go:1472]: > the CI in catalog should use the same code with this PR after merged @wangyikewxgm

Yes, It's true.

Comment (StevenLeiZhang) [reply pkg/addon/addon.go:1472]: > 



> Please make this info clear in the doc. This file is only be used for generating the index.yaml for chart-museum automatically .It's useless when enable an addon. So the addon's maker needn't take aware of the existence of this file and mustn't write this file.

OK. Thanks for reminder.

File: pkg/addon/addon.go (hunk 3 above)
Comment (StevenLeiZhang) [reply pkg/addon/addon.go:1490]: > Is the package's name is {addonName}-{version}.tgz ?

Yes. for example:
```
velaux-v1.3.3.tgz
fluxcd-1.1.0.tgz
```

**Review by wangyikewxgm — APPROVED** [review]
LGTM, Great pr! Thanks.

**Review by Somefive — APPROVED** [review]

**Pull request merged**